README.md
<h1 align="center">
    Opt3001
</h1>

<h4 align="center">
    Python support library for the light sensor OPT3001 from Texas Instruments
</h4>

## Description

## Contents

- [Description](#description)
- [Contents](#contents)
- [Install](#install)
- [Usage](#usage)
- [Flight Manual](#flight-manual)
- [Development](#development)
- [Team](#team)
- [License](#license)

## Install

- sudo apt-get install -y python-smbus
- pip install opt3001

## Usage

```python
import opt3001 from opt3001

address = 0x44

opt = opt3001.OPT3001(address) 

# Configure to run in Continuous conversions mode
opt.write_config_reg(opt3001.I2C_LS_CONFIG_CONT_FULL_800MS)

while(True):
  print(opt.read_lux_float())
  time.sleep(1)
```

### Raspberry pi

Add following line to `/etc/modules` to start i2c module on startup:
``` bash
i2c_dev
```

Run following commands to start the i2c modules manually:
``` bash
$ modprobe i2c-bcm2708
$ modprobe i2c_dev
```

Install `i2c-tools` to test with the terminal

``` bash
$ sudo apt-get install i2c-tools
```

Run following command to get the address of the opt3001

``` bash
$ i2cdetect -y 1

     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- 44 -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --
```

Run the following command to test the opt30001
``` bash
$ i2cget -y 1 0x44 0x00

0x0b
```

## Flight Manual

### Constants

| constant | description | value |
| --- | --- | --- |
| I2C_LS_CONFIG_DEFAULT | OPT3001 default settings  | 0xc810 |
| I2C_LS_CONFIG_CONT_FULL_800MS | settings for continuous measurement | 0xcc10 |
| I2C_LS_REG_RESULT | Result Register | 0x00 |
| I2C_LS_REG_CONFIG | Configuration Register | 0x01 |
| I2C_LS_REG_LOWLIMIT | Low Limit Register | 0x02 |
| I2C_LS_REG_HIGHLIMIT | High Limit Register | 0x03 |
| I2C_LS_REG_MANUFACTURERID | Manufacturer ID | 0x7E |
| I2C_LS_REG_DEVICEID | Device ID Register | 0x7F |


### OPT3001

#### constructor

Input:
- address - i2c address of the opt3001
- bus - default 1

#### read_lux_fixpoint

Read the brightness of the opt3001 with two fixed decimal places

#### read_lux_float

Read the brightness of the opt3001 as float

#### read_manufacture_id

Read manufacture id of the opt3001

#### read_device_id

Read device id of the opt3001

#### write_config_reg

Write data to config register

Input:
- data - configuration data to write to registry

#### read_register_16bit

reads a register of the opt3001

#### write_register_16bit

Write to a register of the opt3001

Input:
- adr - registeradress to write to
- data - 16bit data to write to registry

## Development

### Build
python setup.py sdist bdist_wheel

### Upload
twine upload dist/*

## Team

- Thomas Pöhlmann [(@perryrh0dan)](https://github.com/perryrh0dan)

## Licence

[MIT](https://github.com/perryrh0dan/opt3001/blob/master/license.md)

This repository was generated by [charon](https://github.com/perryrh0dan/charon)


changelog.md
# What's new in opt3001


opt3001/opt3001.py
import smbus

# Constants
I2C_LS_REG_RESULT = 0x00
I2C_LS_REG_CONFIG = 0x01
I2C_LS_REG_LOWLIMIT = 0x02
I2C_LS_REG_HIGHLIMIT = 0x03
I2C_LS_REG_MANUFACTURERID = 0x7E
I2C_LS_REG_DEVICEID = 0x7F

# Configdata for Register "Configuration"
I2C_LS_CONFIG_DEFAULT = 0xc810
# Bit 15..12 Automatic Full-Scale Setting Mode
# Bit 11 Conversion time field: 800ms
# Bit 10..9 Mode of conversion: Shutdown
# Bit 4 Latch field

# Configdata for Register "Configuration"
I2C_LS_CONFIG_CONT_FULL_800MS = 0xcc10
# Bit 15..12 Automatic Full-Scale Setting Mode
# Bit 11 Conversion timefield: 800ms
# 10..9 Mode of conversion: Continuous conversions
# Bit 4 Latch field


class OPT3001:
    def __init__(self, address, bus=1):
        """ -----------------------------------------------
        Name: init

        Description:  initiate the opt3001

        Input:  address - i2c address of the opt3001
                bus - default 1

        Ausgang: opt3001
        """
        self.bus = smbus.SMBus(bus)
        self.address = address

    def read_register_16bit(self, adr):
        """ -----------------------------------------------
        Name: read_register_16bit

        Description: reads a register of the opt3001

        Input:  adr - registeradress to read from

        Ausgang: data
        """
        values = self.bus.read_i2c_block_data(self.address, adr, 2)
        data = (values[0] << 8) | values[1]
        return data

    def write_register_16bit(self, adr, data):
        """ -----------------------------------------------
        Name: write_register_16bit

        Description:  write to a register of the opt3001

        Input:  adr - registeradress to write to
                data - data to write to register

        Ausgang: void
        """
        d1 = (data >> 8)
        d0 = data & 0xFF
        return self.bus.write_i2c_block_data(self.address, adr, [d1, d0])

    def write_config_reg(self, data):
        """ -----------------------------------------------
        Name: write_config_req

        Description:  write to config register

        Input:  data - data to write to register

        Ausgang: void
        """
        return self.write_register_16bit(I2C_LS_REG_CONFIG, data)

    def read_manufacture_id(self):
        """ -----------------------------------------------
        Name: read_manufacture_id

        Description:  read manufacture id of the opt3001

        Input: void

        Ausgang: manufacture id
        """
        return self.read_register_16bit(I2C_LS_REG_MANUFACTURERID)

    def read_device_id(self):
        """ -----------------------------------------------
        Name: read_device_id

        Description:  read device id of the opt3001

        Input: void

        Ausgang: manufacture id
        """
        return self.read_register_16bit(I2C_LS_REG_DEVICEID)

    def read_lux_fixpoint(self):
        """ -----------------------------------------------
        Name: read_lux_fixpoint

        Description:  read the brightness of the opt3001 with two fixed decimal places

        Input:  void

        Ausgang: manufacture id
        """
        # Register Value
        req_value = self.read_register_16bit(I2C_LS_REG_RESULT)

        # Convert to LUX
        mantisse = req_value & 0x0fff
        exponent = (req_value & 0xf000) >> 12

        return 2**exponent * mantisse  # mantisse << exponent;

    def read_lux_float(self):
        """ -----------------------------------------------
        Name: read_lux_float

        Description:  read the brightness of the opt3001 as float

        Input:  void

        Ausgang: manufacture id
        """
        # Register Value
        req_value = self.read_register_16bit(I2C_LS_REG_RESULT)

        # Convert to LUX
        mantisse = req_value & 0x0fff
        exponent = (req_value & 0xf000) >> 12

        return 2**exponent * mantisse * 0.01  # mantisse << exponent * 0.01;
